src/rest-server/src/models/v2/job/k8s.js

```diff
@@ -615,6 +649,102 @@ const getSshInfo = async (frameworkName) => {
   throw createError('Not Found', 'NoJobSshInfoError', `SSH info of job ${frameworkName} is not found.`);
 };
 
+const extractDiagnosticsInfo = (diag) => {
+  if (_.isEmpty(diag)) {
+    return null;
+  }
+  // The pattern for diagnostics is "podPattern unmatched: json" or "podPattern nmatched: json"
+  const regex = /matched: (.*)/;
+  const matches = diag.match(regex);
+  if (matches.length < 2) {
+    return null;
+  }
+  try {
+    return JSON.parse(matches[1]);
+  } catch (error) {
+    logger.warn('Get diagnostics info failed', error);
+  }
+  return null;
+};
+
+const extractRuntimeOutput = (diag) => {
+  if (_.isEmpty(diag)) {
+    return null;
+  }
+
+  // Do not container user container info
+  if (_.isNil(diag.containers) || diag.containers.length < 2) {
+    return null;
+  }
+  const message = diag.containers[1].message;
+
+  const anchor1 = /\[PAI_RUNTIME_ERROR_START\]/;
+  const anchor2 = /\[PAI_RUNTIME_ERROR_END\]/;
+  const match1 = message.match(anchor1);
+  const match2 = message.match(anchor2);
+  if (match1 !== null && match2 !== null) {
+    const start = match1.index + match1[0].length;
+    const end = match2.index;
+    const output = message.substring(start, end).trim();
+    return yaml.safeLoad(output);
+  }
+};
+
+const extractContainerOutput = (diag) => {
+  if (_.isEmpty(diag) || _.isEmpty(diag.containers)) {
+    return null;
+  }
+  const anchor = /\[PAI_RUNTIME_ERROR_START\]/;
+
+  const obj = {};
+  diag.containers.forEach((container) => {
+    let message = container.message;
+    if (_.isEmpty(message)) {
+      return;
+    }
+    const match = message.match(anchor);
+    if (match !== null) {
+      message = message.substring(0, match.index);
+    }
```

Comment:
In which cases, the message is not empty?

Reply:
In yarn, we need this because, we use it to catch YarnContainerScripts stderr, in k8s, do we still need?


     a) Get error message from container stderr (ExitCodeException Message)
     ExitCodeException exitCode=127: YarnContainerScripts/0.sh: line 58: [: -gt: unary operator expected
             at org.apache.hadoop.util.Shell.runCommand(Shell.java:998)
             at org.apache.hadoop.util.Shell.run(Shell.java:884)
             at org.apache.hadoop.util.Shell$ShellCommandExecutor.execute


---
In reply to: [321554938](https://github.com/microsoft/pai/pull/3575#discussion_r321554938) [](ancestors = 321554938)

Reply:
If all the message here is extracted from pod `termination message`. 
Then the message will be empty.
I'm not sure if controller customize the message here. If not, will remove this function.
And `appExitMessages.container` will be null in k8s version

Reply:
If we do not need this. Will leave it as null for API compatibility

Reply:
it is raw or substring of termination message, so let's remove it

---
In reply to: [321557904](https://github.com/microsoft/pai/pull/3575#discussion_r321557904) [](ancestors = 321557904)